### PR TITLE
Add the option to deploy on knative

### PR DESCRIPTION
**Pull request by xiezhang7**
Added the option to deploy the workflow CR with knative.

### CODE REVIEW

**Comment by xiezhang7** [comment]
Hi @ruromero , could you please review this pull request?

I changed some of the `serverless-workflow-operator` to `workflow-operator`.

It seems some versions of knative don't work with images from `quay.io`. I finally managed to get the image `quay.io/ruben/workflow-service:latest` working by re-tagging it and pushing it to `docker.io`.

Thanks.

File: serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go [hunk 1]
```diff
@@ -320,3 +343,58 @@ func getImage(cr *v1.Workflow) string {
 func intPtr(x int32) *int32 {
 	return &x
 }
+
+func newKnativeService(cr *v1.Workflow) *knsv1alpha1.Service {
```
Comment (ruromero) [inline serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go]:
```suggestion
func newKnativeService(cr *v1.Workflow) *knsv1.Service {
```

File: serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go [hunk 2]
```diff
@@ -320,3 +343,58 @@ func getImage(cr *v1.Workflow) string {
...
+	return &knsv1alpha1.Service{
```
Comment (ruromero) [inline serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go]:
```suggestion
	return &knsv1.Service{
```

**Comment by xiezhang7** [comment]
> No need to use v1alpha1

Changed to use v1. Thanks.

File: serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go [hunk 3]
```diff
@@ -320,3 +342,56 @@ func getImage(cr *v1.Workflow) string {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+			Labels:    getLabels(cr),
+		},
+		Spec: knsv1.ServiceSpec{
+			ConfigurationSpec: knsv1.ConfigurationSpec{
+				Template: knsv1.RevisionTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cr.Name + "-v1",
```
Comment (ruromero) [inline serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go]: Why -v1?

File: serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go [hunk 4]
```diff
@@ -320,3 +342,56 @@ func getImage(cr *v1.Workflow) string {
...
+						Namespace: cr.Namespace,
+						Labels:    getDeploymentLabels(cr),
+					},
+					Spec: knsv1.RevisionSpec{
+						PodSpec: corev1.PodSpec{
```
Comment (ruromero) [inline serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go]: You can create a common PodSpec and add it there and also to the v1.Service

**Comment by ruromero** [comment]
One last thing. Can you change the default image to be quay.io/rhsyseng/workflow-service and the deploy/operator.yml to also point to rhsyseng repository?
Thanks

File: serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go (hunk 3 above)
Comment (xiezhang7) [reply serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go]: It is the revision of the revision template. At least, v1alpha1 requires the revision in the revision template name.
But it is not mandatory to assign a name to the revision template. I'll remove the name, and a revision will be generated by knative serving.

File: serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go (hunk 4 above)
Comment (xiezhang7) [reply serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go]: will do

File: serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go (hunk 3 above)
Comment (ruromero) [reply serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go]: Exactly. KNative will create a revision accordingly

File: serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go (hunk 4 above)
Comment (xiezhang7) [reply serverless-workflow-operator/pkg/controller/workflow/workflow_controller.go]: Knative doesn't support env with downward api (https://github.com/knative/serving/issues/4190).
I'll use value `cr.Namespace` for the `NAMESPACE` env.

**Comment by xiezhang7** [comment]
> One last thing. Can you change the default image to be quay.io/rhsyseng/workflow-service and the deploy/operator.yml to also point to rhsyseng repository?
> Thanks

I have changed the images. Thanks.

**Pull request merged**